platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/internal/GradleVersionResolver.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.wrapper.internal;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.gradle.api.GradleException;
+import org.gradle.api.resources.TextResource;
+import org.gradle.api.resources.TextResourceFactory;
+import org.gradle.internal.exceptions.ResolutionProvider;
+import org.gradle.util.GradleVersion;
+import org.gradle.util.internal.DistributionLocator;
+import org.jspecify.annotations.Nullable;
+
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+public class GradleVersionResolver {
+
+    public static final String LATEST = "latest";
+    public static final String NIGHTLY = "nightly";
+    public static final String RELEASE_NIGHTLY = "release-nightly";
+    public static final String RELEASE_CANDIDATE = "release-candidate";
+    public static final String RELEASE_MILESTONE = "release-milestone";
+
+    public static final List<String> PLACE_HOLDERS = Arrays.asList(LATEST, RELEASE_CANDIDATE, RELEASE_MILESTONE, RELEASE_NIGHTLY, NIGHTLY); // order these from most-to-least stable
+
+    private final TextResource latest;
+    private final TextResource releaseCandidate;
+    private final TextResource releaseMilestone;
+    private final TextResource nightly;
+    private final TextResource releaseNightly;
+
+    @Nullable
+    private GradleVersion gradleVersion;
+    private String gradleVersionString = GradleVersion.current().getVersion();
+
+
+    public GradleVersionResolver(TextResourceFactory textFactory) {
+        String versionUrl = DistributionLocator.getBaseUrl() + "/versions";
+        this.latest = textFactory.fromUri(versionUrl + "/current");
+        this.releaseCandidate = textFactory.fromUri(versionUrl + "/release-candidate");
+        this.releaseMilestone = textFactory.fromUri(versionUrl + "/milestone");
+        this.nightly = textFactory.fromUri(versionUrl + "/nightly");
+        this.releaseNightly = textFactory.fromUri(versionUrl + "/release-nightly");
```

Comment:
🤔 Seems like it might be worth building a higher-level abstraction for each type of version.  Perhaps a nested static `enum class` that would contain the string, the ordering, and methods for getting the `TextResource `for each?